File: src/detector_bajas.py
"""
detector_bajas.py

Compara el último run válido contra el run válido anterior.
NO mezcla snapshots de corridas distintas ni de runs inválidos.

Clasificaciones de cambio:
  nuevo                   → item aparece en run actual pero no en el anterior
  desaparecido_no_confirmado → item estaba y ya no aparece
  vendido_confirmado      → sold_quantity subió (señal más fuerte)
  vendido_probable        → available_quantity bajó y el item sigue activo
  status_cambio           → cambió el campo status
  precio_cambio           → cambió el precio más de un 5%
  stock_cambio            → cambió available_quantity sin venta confirmada

Salida:
  0 → detector completó (aunque no haya cambios)
  1 → error técnico o insuficientes runs válidos
"""

import os
import sys
from datetime import datetime
from supabase import create_client

# ─── Config ────────────────────────────────────────────────────────────────────
SUPABASE_URL = os.getenv('SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_KEY')

if not SUPABASE_URL or not SUPABASE_KEY:
    print('❌ FATAL: Faltan SUPABASE_URL o SUPABASE_KEY')
    sys.exit(1)

supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

SEP = '═' * 70
PRECIO_MIN_PCT = 5.0      # mínimo % de cambio de precio para reportar


# ─── Obtener los 2 últimos runs válidos ────────────────────────────────────────
def get_last_two_valid_runs():
    print('📊 Consultando últimos 2 runs válidos en monitor_runs...')
    resp = (
        supabase.table('monitor_runs')
        .select('*')
        .eq('status', 'valid')
        .order('finished_at', desc=True)
        .limit(2)
        .execute()
    )
    data = resp.data or []
    if len(data) < 2:
        print(f'⚠️  Solo {len(data)} run(s) válido(s). Se necesitan al menos 2 para comparar.')
        return None, None
    current  = data[0]   # el más reciente
    previous = data[1]   # el inmediatamente anterior válido
    print(f'  Run actual:   {current["run_id"]}  ({current["total_items"]} items,'
          f' {current["finished_at"][:19]})')
    print(f'  Run anterior: {previous["run_id"]} ({previous["total_items"]} items,'
          f' {previous["finished_at"][:19]})')
    return current, previous


# ─── Obtener todos los snapshots de un run (indexados por item_id) ─────────────
def get_snapshots_for_run(run_id):
    resp = (
        supabase.table('snapshots')
        .select('*')
        .eq('run_id', run_id)
        .execute()
    )
    data = resp.data or []
    by_item = {}
    for snap in data:
        iid = snap.get('item_id') or snap.get('meli_item_id')
        if iid and iid not in by_item:
            by_item[iid] = snap
    return by_item


# ─── Verificar si ya se detectó para este run (evitar duplicados) ──────────────
def already_detected(current_run_id):
    try:
        resp = (
            supabase.table('bajas_detectadas')
            .select('id')
            .eq('detection_run_id', current_run_id)
            .limit(1)
            .execute()
        )
        return len(resp.data or []) > 0
    except Exception:
        # Si la columna no existe todavía, proceder igual
        return False


# ─── Comparar dos sets de snapshots y clasificar cambios ──────────────────────
def detect_changes(current_snaps, previous_snaps, current_run_id):
    cambios = []
    now = datetime.now().isoformat()

    current_ids  = set(current_snaps.keys())
    previous_ids = set(previous_snaps.keys())

    # ── NUEVOS ────────────────────────────────────────────────────────────────
    for iid in current_ids - previous_ids:
        snap = current_snaps[iid]
        cambios.append({
            'tipo':             'nuevo',
            'item_id':          iid,
            'meli_item_id':     snap.get('meli_item_id', iid),
            'seller_id':        snap.get('seller_id'),
            'title':            snap.get('title'),
            'precio_nuevo':     snap.get('price'),
            'status_nuevo':     snap.get('status'),
            'detection_run_id': current_run_id,
            'fecha_deteccion':  now,
        })

    # ── DESAPARECIDOS ─────────────────────────────────────────────────────────
    for iid in previous_ids - current_ids:
        snap = previous_snaps[iid]
        cambios.append({
            'tipo':             'desaparecido_no_confirmado',
            'item_id':          iid,
            'meli_item_id':     snap.get('meli_item_id', iid),
            'seller_id':        snap.get('seller_id'),
            'title':            snap.get('title'),
            'precio_anterior':  snap.get('price'),
            'status_anterior':  snap.get('status'),
            'detection_run_id': current_run_id,
            'fecha_deteccion':  now,
        })

    # ── ITEMS EN AMBOS RUNS ───────────────────────────────────────────────────
    for iid in current_ids & previous_ids:
        curr = current_snaps[iid]
        prev = previous_snaps[iid]

        sold_curr  = curr.get('sold_quantity')  or 0
        sold_prev  = prev.get('sold_quantity')  or 0
        avail_curr = curr.get('available_quantity') or 0
        avail_prev = prev.get('available_quantity') or 0
        price_curr = curr.get('price') or 0
        price_prev = prev.get('price') or 0
        status_curr = curr.get('status') or ''
        status_prev = prev.get('status') or ''

        base = {
            'item_id':          iid,
            'meli_item_id':     curr.get('meli_item_id', iid),
            'seller_id':        curr.get('seller_id'),
            'title':            curr.get('title'),
            'precio_anterior':  price_prev,
            'precio_nuevo':     price_curr,
            'detection_run_id': current_run_id,
            'fecha_deteccion':  now,
        }

        # 1. vendido_confirmado: sold_quantity subió (señal más fuerte, cortar análisis)
        if sold_curr > sold_prev and sold_curr > 0:
            cambios.append({**base,
                'tipo':             'vendido_confirmado',
                'unidades_vendidas': sold_curr - sold_prev,
                'stock_anterior':   avail_prev,
                'stock_nuevo':      avail_curr,
            })
            continue  # señal más fuerte para este item

        # 2. vendido_probable: available_quantity bajó y el item sigue activo
        vendido_probable = avail_curr < avail_prev and status_curr == 'active'
        if vendido_probable:
            cambios.append({**base,
                'tipo':           'vendido_probable',
                'stock_anterior': avail_prev,
                'stock_nuevo':    avail_curr,
            })
            # no cortamos: podría haber también status o precio cambiado

        # 3. status_cambio
        if status_curr != status_prev:
            cambios.append({**base,
                'tipo':            'status_cambio',
                'status_anterior': status_prev,
                'status_nuevo':    status_curr,
            })

        # 4. precio_cambio (>= 5%)
        if price_prev > 0 and price_curr > 0:
            pct = (price_curr - price_prev) / price_prev * 100
            if abs(pct) >= PRECIO_MIN_PCT:
                cambios.append({**base,
                    'tipo':              'precio_cambio',
                    'cambio_porcentaje': round(pct, 2),
                })

        # 5. stock_cambio (sin venta confirmada ni probable)
        elif avail_curr != avail_prev and not vendido_probable:
            cambios.append({**base,
                'tipo':           'stock_cambio',
                'stock_anterior': avail_prev,
                'stock_nuevo':    avail_curr,
            })

    return cambios


# ─── Guardar cambios en bajas_detectadas ──────────────────────────────────────
def guardar_cambios(cambios):
    if not cambios:
        print('ℹ️  Sin cambios para guardar')
        return 0

    print(f'\n💾 Insertando {len(cambios)} cambios en bajas_detectadas...')

    rows = []
    for c in cambios:
        rows.append({
            'seller_id':        c.get('seller_id'),
            'item_id':          c.get('item_id'),
            'meli_item_id':     c.get('meli_item_id'),
            'title':            c.get('title'),
            'tipo':             c.get('tipo'),
            'precio_anterior':  c.get('precio_anterior'),
            'precio_nuevo':     c.get('precio_nuevo'),
            'cambio_porcentaje':c.get('cambio_porcentaje'),
            'status_anterior':  c.get('status_anterior'),
            'status_nuevo':     c.get('status_nuevo'),
            'unidades_vendidas':c.get('unidades_vendidas'),
            'stock_anterior':   c.get('stock_anterior'),
            'stock_nuevo':      c.get('stock_nuevo'),
            'detection_run_id': c.get('detection_run_id'),
            'fecha_deteccion':  c.get('fecha_deteccion'),
        })

    try:
        supabase.table('bajas_detectadas').insert(rows).execute()
        print(f'✅ {len(rows)} cambios guardados')
        return len(rows)
    except Exception as e:
        print(f'⚠️  Bulk insert falló ({e}), intentando de a uno...')
        ok = 0
        for row in rows:
            try:
                supabase.table('bajas_detectadas').insert([row]).execute()
                ok += 1
            except Exception as e2:
                print(f'  ❌ {row.get("item_id")} [{row.get("tipo")}]: {e2}')
        return ok


# ─── Main ──────────────────────────────────────────────────────────────────────
def main():
    print(f'\n{SEP}')
    print('DETECTOR DE CAMBIOS MLU — INICIO')
    print(f'Ejecutado: {datetime.now().isoformat()}')
    print(f'{SEP}\n')

    current_run, previous_run = get_last_two_valid_runs()

    if not current_run or not previous_run:
        print('ℹ️  Sin suficientes runs válidos para comparar — saliendo sin error')
        sys.exit(0)

    current_run_id  = current_run['run_id']
    previous_run_id = previous_run['run_id']

    # Evitar duplicados por misma corrida
    if already_detected(current_run_id):
        print(f'⚠️  Ya se detectaron cambios para {current_run_id} — saltando')
        sys.exit(0)

    # Cargar snapshots
    print(f'\n📥 Snapshots del run actual ({current_run_id})...')
    current_snaps = get_snapshots_for_run(current_run_id)
    print(f'   {len(current_snaps)} items únicos')

    print(f'\n📥 Snapshots del run anterior ({previous_run_id})...')
    previous_snaps = get_snapshots_for_run(previous_run_id)
    print(f'   {len(previous_snaps)} items únicos')

    # Comparar
    print('\n🔍 Comparando...\n')
    cambios = detect_changes(current_snaps, previous_snaps, current_run_id)

    # Resumen
    print(f'\n{SEP}')
    print('RESUMEN DE CAMBIOS DETECTADOS')
    print(SEP)
    tipos = {}
    for c in cambios:
        tipos[c['tipo']] = tipos.get(c['tipo'], 0) + 1

    if tipos:
        for tipo in sorted(tipos):
            print(f'  {tipo:35}: {tipos[tipo]:4}')
        print(f'  {"TOTAL":35}: {len(cambios):4}')
    else:
        print('  Sin cambios detectados entre los dos runs')

    print(SEP)

    # Detalle
    if cambios:
        print('\nDetalle:')
        for c in cambios:
            title = (c.get('title') or '')[:45]
            extra = ''
            if c['tipo'] == 'vendido_confirmado':
                extra = f' | +{c.get("unidades_vendidas", "?")} unid'
            elif c['tipo'] == 'precio_cambio':
                extra = f' | {c.get("cambio_porcentaje", 0):+.1f}%'
            elif c['tipo'] in ('vendido_probable', 'stock_cambio'):
                extra = f' | stock {c.get("stock_anterior")}→{c.get("stock_nuevo")}'
            elif c['tipo'] in ('status_cambio',):
                extra = f' | {c.get("status_anterior")}→{c.get("status_nuevo")}'
            print(f'  [{c["tipo"]}] {title}{extra}')

    # Guardar
    insertados = guardar_cambios(cambios)

    print(f'\n{SEP}')
    print(f'✅ DETECTOR COMPLETADO — {insertados} cambios guardados')
    print(f'{SEP}\n')


if __name__ == '__main__':
    main()

File: src/detector_bajas.js
'use strict';

const fs = require('fs');

/**
 * detector_bajas.js — v5, mínimo y auditable
 *
 * Por cada seller activo, de forma independiente:
 *   1. Encuentra los 2 run_ids más recientes con datos en snapshots
 *   2. Construye sets de item_ids para cada run
 *   3. Calcula desaparecidos (set_anterior por diseño UPSERT)
 *   4. Loguea todo lo suficiente para verificar a mano
 *   5. Graba en bajas_detectadas (con chequeo de idempotencia por seller)
 *
 * NOTA sobre el diseño UPSERT de snapshots:
 *   Cada item tiene UNA fila con el run_id del último scrape donde apareció.
 *   Por tanto, para seller X:
 *     set_actual   = WHERE seller_id=X AND run_id=run_actual   → activos ahora
 *     set_anterior = WHERE seller_id=X AND run_id=run_anterior → desaparecidos
 *   Los sets son DISJUNTOS por construcción — no hace falta restar.
 *
 * Detección de "nuevos" desactivada en esta fase.
 *
 * Salida:
 *   0 → completó (aunque sin cambios)
 *   1 → error técnico
 */

['SUPABASE_URL', 'SUPABASE_KEY'].forEach(k => {
  if (!process.env[k]) { console.error(`FATAL: falta variable ${k}`); process.exit(1); }
});

const https    = require('https');
const { createClient } = require('@supabase/supabase-js');
const supabase = createClient(process.env.SUPABASE_URL, process.env.SUPABASE_KEY);

const OXYLABS_USER = process.env.OXYLABS_USER || '';
const OXYLABS_PASS = process.env.OXYLABS_PASS || '';

// ─── Enriquecer items: título + precio via Oxylabs → ML API ──────────────────
async function enrichItems(itemIds) {
  const details = {};
  if (!itemIds.length || !OXYLABS_USER) return details;

  const CHUNK = 20;
  for (let i = 0; i < itemIds.length; i += CHUNK) {
    const chunk   = itemIds.slice(i, i + CHUNK);
    const mlUrl   = `https://api.mercadolibre.com/items?ids=${chunk.join(',')}`;
    const payload = JSON.stringify({ source: 'universal', url: mlUrl });
    const auth    = Buffer.from(`${OXYLABS_USER}:${OXYLABS_PASS}`).toString('base64');

    try {
      const raw = await new Promise((resolve, reject) => {
        const req = https.request({
          hostname: 'realtime.oxylabs.io', path: '/v1/queries', method: 'POST',
          headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(payload), 'Authorization': 'Basic ' + auth },
        }, res => { const c = []; res.on('data', x => c.push(x)); res.on('end', () => resolve(Buffer.concat(c).toString())); });
        req.on('error', reject);
        req.setTimeout(30000, () => { req.destroy(); reject(new Error('timeout')); });
        req.write(payload); req.end();
      });

      const oxy  = JSON.parse(raw);
      let content = oxy.results?.[0]?.content;
      if (typeof content === 'string') content = JSON.parse(content);
      if (!Array.isArray(content)) continue;

      for (const row of content) {
        const b = row.body || row;
        if (b?.id) {
          details[b.id] = {
            title:     b.title     || null,
            price:     b.price     ?? null,
            currency:  b.currency_id || 'UYU',
            thumbnail: b.thumbnail || b.pictures?.[0]?.url || null,
          };
        }
      }
    } catch (e) {
      console.warn(`  ⚠️  enrichItems chunk error: ${e.message}`);
    }
  }
  console.log(`  🔍 enriquecidos: ${Object.keys(details).length}/${itemIds.length}`);
  return details;
}

const SEP  = '═'.repeat(70);
const SEP2 = '─'.repeat(70);

// ─── Sellers activos ──────────────────────────────────────────────────────────
async function getActiveSellers() {
  const { data, error } = await supabase
    .from('sellers')
    .select('seller_id, nombre_real, nickname')
    .eq('activo', true);
  if (error) throw new Error(`getActiveSellers: ${error.message}`);
  return data || [];
}

// ─── 2 run_ids más recientes para un seller, con timestamps ──────────────────
async function getLastTwoRunIds(sellerId) {
  const { data, error } = await supabase
    .from('snapshots')
    .select('run_id, checked_at')
    .eq('seller_id', String(sellerId))
    .order('checked_at', { ascending: false })
    .limit(5000);

  if (error) throw new Error(`getLastTwoRunIds(${sellerId}): ${error.message}`);

  const seen = [];
  const ts   = {};
  for (const row of (data || [])) {
    if (row.run_id && !seen.includes(row.run_id)) {
      seen.push(row.run_id);
      ts[row.run_id] = row.checked_at;
    }
    if (seen.length >= 2) break;
  }
  return { runIds: seen, ts };
}

// ─── Items de un seller en un run concreto (paginado, deduplicado) ────────────
async function getItemSet(sellerId, runId) {
  const items    = new Set();
  const pageSize = 1000;
  let   offset   = 0;

  while (true) {
    const { data, error } = await supabase
      .from('snapshots')
      .select('item_id, meli_item_id')
      .eq('seller_id', String(sellerId))
      .eq('run_id', runId)
      .range(offset, offset + pageSize - 1);

    if (error) throw new Error(`getItemSet(${sellerId}, ${runId}): ${error.message}`);
    const batch = data || [];
    for (const row of batch) {
      const id = row.item_id || row.meli_item_id;
      if (id) items.add(id);
    }
    if (batch.length < pageSize) break;
    offset += pageSize;
  }
  return items;
}

// ─── Idempotencia: ¿ya guardamos desaparecidos para este seller en este par de runs? ──
// Verifica si alguno de los items desaparecidos ya existe en bajas_detectadas
// para este seller con tipo='desaparecido_no_confirmado'.
async function alreadyDetectedSeller(sellerId, desaparecidos) {
  if (!desaparecidos.length) return false;
  const sample = desaparecidos.slice(0, 5);
  const { data, error } = await supabase
    .from('bajas_detectadas')
    .select('item_id')
    .eq('seller_id', String(sellerId))
    .eq('tipo', 'desaparecido_no_confirmado')
    .in('item_id', sample)
    .limit(1);
  if (error) return false; // si no se puede verificar, dejamos pasar
  return (data?.length ?? 0) > 0;
}

// ─── Procesar un seller ────────────────────────────────────────────────────────
async function processSeller(seller) {
  const sellerId = String(seller.seller_id);
  const name     = seller.nombre_real || seller.nickname || sellerId;

  console.log(`\n${SEP2}`);
  console.log(`Seller: ${name}  (id: ${sellerId})`);

  // 1. Runs
  const { runIds, ts } = await getLastTwoRunIds(sellerId);

  if (runIds.length < 2) {
    console.log(`  ⏭️  Solo ${runIds.length} run(s) — se necesitan 2 para comparar. Salteando.`);
    return null;
  }

  const [runActual, runAnterior] = runIds;
  console.log(`  run_actual:   ${runActual}  (${(ts[runActual]  || '').slice(0, 19)})`);
  console.log(`  run_anterior: ${runAnterior}  (${(ts[runAnterior] || '').slice(0, 19)})`);

  // 2. Sets de items
  const setActual   = await getItemSet(sellerId, runActual);
  const setAnterior = await getItemSet(sellerId, runAnterior);

  console.log(`  items run_actual:   ${setActual.size}`);
  console.log(`  items run_anterior: ${setAnterior.size}  ← estos son los desaparecidos (UPSERT)`);

  // 3. Desaparecidos (BAJADAS)
  // Por diseño UPSERT: setAnterior = items que ya no están en setActual → son las bajas.
  // "nuevos" desactivado: con UPSERT los sets son disjuntos → falsos positivos masivos.
  const desaparecidos = [...setAnterior];

  console.log(`  desaparecidos: ${desaparecidos.length}`);

  // 4. Muestra auditable
  if (desaparecidos.length > 0) {
    const muestra = desaparecidos.slice(0, 8).join(', ');
    console.log(`  muestra desaparecidos: ${muestra}${desaparecidos.length > 8 ? ` ... (${desaparecidos.length - 8} más)` : ''}`);
  }

  return { sellerId, name, runActual, runAnterior, setActual, setAnterior, desaparecidos, nuevos: [] };
}

// ─── Guardar en bajas_detectadas ──────────────────────────────────────────────
async function saveChanges(rows) {
  if (!rows.length) return 0;

  const { error } = await supabase.from('bajas_detectadas').insert(rows);
  if (!error) return rows.length;

  console.warn(`  ⚠️  Bulk insert falló (${error.message}), insertando de a uno...`);
  let ok = 0;
  for (const row of rows) {
    const { error: e2 } = await supabase.from('bajas_detectadas').insert([row]);
    if (!e2) ok++;
    else console.warn(`    ❌ ${row.item_id} [${row.tipo}]: ${e2.message}`);
  }
  return ok;
}

// ─── Main ─────────────────────────────────────────────────────────────────────
(async () => {
  const DET_RUN_ID = `det_${Date.now()}`;
  const NOW        = new Date().toISOString();

  console.log(`\n${SEP}`);
  console.log('DETECTOR BAJAS v4 — INICIO');
  console.log(`det_run_id: ${DET_RUN_ID}`);
  console.log(`timestamp:  ${NOW}`);
  console.log(`${SEP}`);

  let sellers;
  try {
    sellers = await getActiveSellers();
    console.log(`\nSellers activos: ${sellers.length}`);
  } catch (e) {
    console.error(`❌ ${e.message}`);
    process.exit(1);
  }

  const results = [];

  for (const seller of sellers) {
    try {
      const r = await processSeller(seller);
      if (r) results.push(r);
    } catch (e) {
      console.error(`  ❌ Error en seller ${seller.seller_id}: ${e.message}`);
    }
  }

  // ── Resumen tabular ──────────────────────────────────────────────────────────
  console.log(`\n${SEP}`);
  console.log('RESUMEN POR SELLER');
  console.log(SEP);
  console.log(
    'Seller'.padEnd(38) + ' | ' +
    'ant'.padStart(4)   + ' | ' +
    'act'.padStart(4)   + ' | ' +
    'desap'.padStart(5)
  );
  console.log(SEP2);

  let totDes = 0;
  let totNuevos = 0;
  for (const r of results) {
    console.log(
      r.name.slice(0, 37).padEnd(38) + ' | ' +
      String(r.setAnterior.size).padStart(4) + ' | ' +
      String(r.setActual.size).padStart(4)   + ' | ' +
      String(r.desaparecidos.length).padStart(5)
    );
    totDes += r.desaparecidos.length;
    totNuevos += r.nuevos.length;
  }
  console.log(SEP2);
  console.log(
    'TOTAL'.padEnd(38) + ' | ' +
    '    '             + ' | ' +
    '    '             + ' | ' +
    String(totDes).padStart(5)
  );
  console.log(SEP);

  // ── Guardar (con idempotencia por seller) ────────────────────────────────────
  let totalSaved   = 0;
  let totalSkipped = 0;

  for (const r of results) {
    if (!r.desaparecidos.length) continue;

    // Chequeo de idempotencia: si ya existen entradas para este seller+items → saltar
    const alreadyDone = await alreadyDetectedSeller(r.sellerId, r.desaparecidos);
    if (alreadyDone) {
      console.log(`  ⏭️  ${r.name}: detección ya guardada, salteando.`);
      totalSkipped++;
      continue;
    }

    // Enriquecer con título y precio antes de guardar
    const enriched = await enrichItems(r.desaparecidos);

    const rows = r.desaparecidos.map(id => ({
      seller_id:     r.sellerId,
      item_id:       id,
      meli_item_id:  id,
      tipo:          'desaparecido_no_confirmado',
      run_id:        DET_RUN_ID,
      fecha_deteccion: NOW,
      title:         enriched[id]?.title     ?? null,
      price_anterior: enriched[id]?.price    ?? null,
    }));

    console.log(`\nGuardando ${r.desaparecidos.length} bajas para ${r.name}...`);
    const saved = await saveChanges(rows);
    console.log(`  ✅ ${saved}/${rows.length} guardados`);
    totalSaved += saved;
  }

  // ── Escribir summary JSON para write_sheets.js ───────────────────────────────
  const summary = {
    det_run_id: DET_RUN_ID,
    timestamp:  NOW,
    sellers: results.map(r => ({
      seller_id:      r.sellerId,
      name:           r.name,
      run_actual:     r.runActual,
      run_anterior:   r.runAnterior,
      items_actuales: r.setActual.size,
      desaparecidos:  r.desaparecidos.length,
    })),
  };
  fs.mkdirSync('output', { recursive: true });
  fs.writeFileSync('output/detector_summary.json', JSON.stringify(summary, null, 2));
  console.log('\n📄 Summary escrito en output/detector_summary.json');

  // ── Cierre ───────────────────────────────────────────────────────────────────
  console.log(`\n${SEP}`);
  console.log('DETECTOR COMPLETADO');
  console.log(`  Sellers procesados: ${results.length}/${sellers.length}`);
  console.log(`  Desaparecidos:      ${totDes}`);
  console.log(`  Guardados:          ${totalSaved}`);
  console.log(`  Salteados (ya guardados): ${totalSkipped}`);
  console.log(`${SEP}\n`);

  process.exit(0);
})();
